stac_fastapi/pgstac/stac_fastapi/pgstac/core.py

```diff
@@ -0,0 +1,312 @@
+"""Item crud client."""
+import re
+from datetime import datetime
+from typing import Any, Dict, List, Optional, Union
+from urllib.parse import urljoin
+
+import attr
+import orjson
+from buildpg import render
+from fastapi.responses import ORJSONResponse
+from stac_pydantic import Collection, Item, ItemCollection
+from stac_pydantic.api import ConformanceClasses, LandingPage
+from stac_pydantic.shared import Link, MimeTypes, Relations
+
+from stac_fastapi.pgstac.models.links import CollectionLinks, ItemLinks, PagingLinks
+from stac_fastapi.pgstac.types.search import PgstacSearch
+from stac_fastapi.types.core import BaseCoreClient
+from stac_fastapi.types.errors import NotFoundError
+
+NumType = Union[float, int]
+
+
+@attr.s
+class CoreCrudClient(BaseCoreClient):
+    """Client for core endpoints defined by stac."""
+
+    async def landing_page(self, **kwargs) -> ORJSONResponse:
+        """Landing page.
+
+        Called with `GET /`.
+
+        Returns:
+            API landing page, serving as an entry point to the API.
+        """
+        request = kwargs["request"]
+        base_url = str(request.base_url)
+        landing_page = LandingPage(
```

Comment:
Missing the "collections" link https://github.com/stac-utils/stac-fastapi/blob/f6a9007e74691143110fe0d3ec947bede024b99c/stac_fastapi/types/stac_fastapi/types/core.py#L148-L152

Perhaps we can define a `def get_landing_page(self)` method that both implementations could use, to ensure consistency?

Reply:
@lossyrob get_landing_page is still dependent on the backend (to get the list of collections), so the direct get_landing_page() still needs to be defined for both backends. I do think that we should look to move the Link generation code into a common place across backends, but I think it's best to do that after this has been merged so we can track things a bit more granularly that are touching both backends. For this PR, I've tried to avoid doing too many things that would change/affect the sqlalchemy backend.

I just pushed a commit that:

1. moves the test data out of sqlalchemy
2. modifies ingest_joplin.py to take an app_host parameter
3. adds ingest_joplin.py runner to docker-compose for pgstac backend 
4. modifies /collections to return empty array rather than not found error (and changes test that was expecting not found)
5. adds missing collections/ link to landing page

Reply:
Hitting http://localhost:8082/collections/joplin/items with the development data, looks like the `null` issue has popped up again?

<details>
  <summary>Response JSON</summary>

```javascript
{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -94.6334839,
              37.0595608
            ],
            [
              -94.6334839,
              37.0332547
            ],
            [
              -94.6005249,
              37.0332547
            ],
            [
              -94.6005249,
              37.0595608
            ],
            [
              -94.6334839,
              37.0595608
            ]
          ]
        ]
      },
      "properties": {
        "gsd": 0.5971642834779395,
        "title": null,
        "width": 2500,
        "height": 2500,
        "created": null,
        "mission": null,
        "updated": null,
        "datetime": "2000-02-02T00:00:00Z",
        "platform": null,
        "proj:epsg": 3857,
        "providers": null,
        "description": null,
        "instruments": null,
        "orientation": "nadir",
        "end_datetime": null,
        "constellation": null,
        "start_datetime": null
      },
      "id": "fe916452-ba6f-4631-9154-c249924a122d",
      "bbox": [
        -94.6334839,
        37.0332547,
        -94.6005249,
        37.0595608
      ],
      "stac_version": "1.0.0-beta.2",
      "assets": {
        "COG": {
          "gsd": null,
          "href": "https://arturo-stac-api-test-data.s3.amazonaws.com/joplin/images/may24C355000e4102500n.tif",
          "type": "image/tiff; application=geotiff; profile=cloud-optimized",
          "bands": null,
          "roles": null,
          "title": "NOAA STORM COG",
          "created": null,
          "mission": null,
          "updated": null,
          "platform": null,
          "multihash": null,
          "providers": null,
          "description": null,
          "instruments": null,
          "end_datetime": null,
          "constellation": null,
          "polarizations": null,
          "start_datetime": null
        }
      },
      "links": [
        {
          "href": "http://localhost:8082/collections/joplin",
          "rel": "collection",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/collections/joplin",
          "rel": "parent",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/",
          "rel": "root",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/collections/joplin/items/fe916452-ba6f-4631-9154-c249924a122d",
          "rel": "self",
          "type": "application/geo+json"
        },
        {
          "href": "http://localhost:8082/collections/joplin/items/fe916452-ba6f-4631-9154-c249924a122d/tiles",
          "rel": "alternate",
          "type": "application/json",
          "title": "tiles"
        }
      ],
      "collection": "joplin"
    },
    {
      "type": "Feature",
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -94.6911621,
              37.1033841
            ],
            [
              -94.6911621,
              37.0770932
            ],
            [
              -94.6582031,
              37.0770932
            ],
            [
              -94.6582031,
              37.1033841
            ],
            [
              -94.6911621,
              37.1033841
            ]
          ]
        ]
      },
      "properties": {
        "gsd": 0.5971642834779395,
        "title": null,
        "width": 2500,
        "height": 2500,
        "created": null,
        "mission": null,
        "updated": null,
        "datetime": "2000-02-02T00:00:00Z",
        "platform": null,
        "proj:epsg": 3857,
        "providers": null,
        "description": null,
        "instruments": null,
        "orientation": "nadir",
        "end_datetime": null,
        "constellation": null,
        "start_datetime": null
      },
      "id": "f7f164c9-cfdf-436d-a3f0-69864c38ba2a",
      "bbox": [
        -94.6911621,
        37.0770932,
        -94.6582031,
        37.1033841
      ],
      "stac_version": "1.0.0-beta.2",
      "assets": {
        "COG": {
          "gsd": null,
          "href": "https://arturo-stac-api-test-data.s3.amazonaws.com/joplin/images/may24C350000e4107500n.tif",
          "type": "image/tiff; application=geotiff; profile=cloud-optimized",
          "bands": null,
          "roles": null,
          "title": "NOAA STORM COG",
          "created": null,
          "mission": null,
          "updated": null,
          "platform": null,
          "multihash": null,
          "providers": null,
          "description": null,
          "instruments": null,
          "end_datetime": null,
          "constellation": null,
          "polarizations": null,
          "start_datetime": null
        }
      },
      "links": [
        {
          "href": "http://localhost:8082/collections/joplin",
          "rel": "collection",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/collections/joplin",
          "rel": "parent",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/",
          "rel": "root",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/collections/joplin/items/f7f164c9-cfdf-436d-a3f0-69864c38ba2a",
          "rel": "self",
          "type": "application/geo+json"
        },
        {
          "href": "http://localhost:8082/collections/joplin/items/f7f164c9-cfdf-436d-a3f0-69864c38ba2a/tiles",
          "rel": "alternate",
          "type": "application/json",
          "title": "tiles"
        }
      ],
      "collection": "joplin"
    },
    {
      "type": "Feature",
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -94.4631958,
              37.0836668
            ],
            [
              -94.4631958,
              37.057369
            ],
            [
              -94.4329834,
              37.057369
            ],
            [
              -94.4329834,
              37.0836668
            ],
            [
              -94.4631958,
              37.0836668
            ]
          ]
        ]
      },
      "properties": {
        "gsd": 0.5971642834779395,
        "title": null,
        "width": 2500,
        "height": 2500,
        "created": null,
        "mission": null,
        "updated": null,
        "datetime": "2000-02-02T00:00:00Z",
        "platform": null,
        "proj:epsg": 3857,
        "providers": null,
        "description": null,
        "instruments": null,
        "orientation": "nadir",
        "end_datetime": null,
        "constellation": null,
        "start_datetime": null
      },
      "id": "f734401c-2df0-4694-a353-cdd3ea760cdc",
      "bbox": [
        -94.4631958,
        37.057369,
        -94.4329834,
        37.0836668
      ],
      "stac_version": "1.0.0-beta.2",
      "assets": {
        "COG": {
          "gsd": null,
          "href": "https://arturo-stac-api-test-data.s3.amazonaws.com/joplin/images/may24C370000e4105000n.tif",
          "type": "image/tiff; application=geotiff; profile=cloud-optimized",
          "bands": null,
          "roles": null,
          "title": "NOAA STORM COG",
          "created": null,
          "mission": null,
          "updated": null,
          "platform": null,
          "multihash": null,
          "providers": null,
          "description": null,
          "instruments": null,
          "end_datetime": null,
          "constellation": null,
          "polarizations": null,
          "start_datetime": null
        }
      },
      "links": [
        {
          "href": "http://localhost:8082/collections/joplin",
          "rel": "collection",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/collections/joplin",
          "rel": "parent",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/",
          "rel": "root",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/collections/joplin/items/f734401c-2df0-4694-a353-cdd3ea760cdc",
          "rel": "self",
          "type": "application/geo+json"
        },
        {
          "href": "http://localhost:8082/collections/joplin/items/f734401c-2df0-4694-a353-cdd3ea760cdc/tiles",
          "rel": "alternate",
          "type": "application/json",
          "title": "tiles"
        }
      ],
      "collection": "joplin"
    },
    {
      "type": "Feature",
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -94.43573,
              37.0836668
            ],
            [
              -94.43573,
              37.0595608
            ],
            [
              -94.402771,
              37.0595608
            ],
            [
              -94.402771,
              37.0836668
            ],
            [
              -94.43573,
              37.0836668
            ]
          ]
        ]
      },
      "properties": {
        "gsd": 0.5971642834779395,
        "title": null,
        "width": 2500,
        "height": 2500,
        "created": null,
        "mission": null,
        "updated": null,
        "datetime": "2000-02-02T00:00:00Z",
        "platform": null,
        "proj:epsg": 3857,
        "providers": null,
        "description": null,
        "instruments": null,
        "orientation": "nadir",
        "end_datetime": null,
        "constellation": null,
        "start_datetime": null
      },
      "id": "f490b7af-0019-45e2-854b-3854d07fd063",
      "bbox": [
        -94.43573,
        37.0595608,
        -94.402771,
        37.0836668
      ],
      "stac_version": "1.0.0-beta.2",
      "assets": {
        "COG": {
          "gsd": null,
          "href": "https://arturo-stac-api-test-data.s3.amazonaws.com/joplin/images/may24C372500e4105000n.tif",
          "type": "image/tiff; application=geotiff; profile=cloud-optimized",
          "bands": null,
          "roles": null,
          "title": "NOAA STORM COG",
          "created": null,
          "mission": null,
          "updated": null,
          "platform": null,
          "multihash": null,
          "providers": null,
          "description": null,
          "instruments": null,
          "end_datetime": null,
          "constellation": null,
          "polarizations": null,
          "start_datetime": null
        }
      },
      "links": [
        {
          "href": "http://localhost:8082/collections/joplin",
          "rel": "collection",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/collections/joplin",
          "rel": "parent",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/",
          "rel": "root",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/collections/joplin/items/f490b7af-0019-45e2-854b-3854d07fd063",
          "rel": "self",
          "type": "application/geo+json"
        },
        {
          "href": "http://localhost:8082/collections/joplin/items/f490b7af-0019-45e2-854b-3854d07fd063/tiles",
          "rel": "alternate",
          "type": "application/json",
          "title": "tiles"
        }
      ],
      "collection": "joplin"
    },
    {
      "type": "Feature",
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -94.6884155,
              37.0595608
            ],
            [
              -94.6884155,
              37.0332547
            ],
            [
              -94.6554565,
              37.0332547
            ],
            [
              -94.6554565,
              37.0595608
            ],
            [
              -94.6884155,
              37.0595608
            ]
          ]
        ]
      },
      "properties": {
        "gsd": 0.5971642834779395,
        "title": null,
        "width": 2500,
        "height": 2500,
        "created": null,
        "mission": null,
        "updated": null,
        "datetime": "2000-02-02T00:00:00Z",
        "platform": null,
        "proj:epsg": 3857,
        "providers": null,
        "description": null,
        "instruments": null,
        "orientation": "nadir",
        "end_datetime": null,
        "constellation": null,
        "start_datetime": null
      },
      "id": "f2cca2a3-288b-4518-8a3e-a4492bb60b08",
      "bbox": [
        -94.6884155,
        37.0332547,
        -94.6554565,
        37.0595608
      ],
      "stac_version": "1.0.0-beta.2",
      "assets": {
        "COG": {
          "gsd": null,
          "href": "https://arturo-stac-api-test-data.s3.amazonaws.com/joplin/images/may24C350000e4102500n.tif",
          "type": "image/tiff; application=geotiff; profile=cloud-optimized",
          "bands": null,
          "roles": null,
          "title": "NOAA STORM COG",
          "created": null,
          "mission": null,
          "updated": null,
          "platform": null,
          "multihash": null,
          "providers": null,
          "description": null,
          "instruments": null,
          "end_datetime": null,
          "constellation": null,
          "polarizations": null,
          "start_datetime": null
        }
      },
      "links": [
        {
          "href": "http://localhost:8082/collections/joplin",
          "rel": "collection",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/collections/joplin",
          "rel": "parent",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/",
          "rel": "root",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/collections/joplin/items/f2cca2a3-288b-4518-8a3e-a4492bb60b08",
          "rel": "self",
          "type": "application/geo+json"
        },
        {
          "href": "http://localhost:8082/collections/joplin/items/f2cca2a3-288b-4518-8a3e-a4492bb60b08/tiles",
          "rel": "alternate",
          "type": "application/json",
          "title": "tiles"
        }
      ],
      "collection": "joplin"
    },
    {
      "type": "Feature",
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -94.6609497,
              37.0595608
            ],
            [
              -94.6609497,
              37.0332547
            ],
            [
              -94.6279907,
              37.0332547
            ],
            [
              -94.6279907,
              37.0595608
            ],
            [
              -94.6609497,
              37.0595608
            ]
          ]
        ]
      },
      "properties": {
        "gsd": 0.5971642834779395,
        "title": null,
        "width": 2500,
        "height": 2500,
        "created": null,
        "mission": null,
        "updated": null,
        "datetime": "2000-02-02T00:00:00Z",
        "platform": null,
        "proj:epsg": 3857,
        "providers": null,
        "description": null,
        "instruments": null,
        "orientation": "nadir",
        "end_datetime": null,
        "constellation": null,
        "start_datetime": null
      },
      "id": "ea0fddf4-56f9-4a16-8a0b-f6b0b123b7cf",
      "bbox": [
        -94.6609497,
        37.0332547,
        -94.6279907,
        37.0595608
      ],
      "stac_version": "1.0.0-beta.2",
      "assets": {
        "COG": {
          "gsd": null,
          "href": "https://arturo-stac-api-test-data.s3.amazonaws.com/joplin/images/may24C352500e4102500n.tif",
          "type": "image/tiff; application=geotiff; profile=cloud-optimized",
          "bands": null,
          "roles": null,
          "title": "NOAA STORM COG",
          "created": null,
          "mission": null,
          "updated": null,
          "platform": null,
          "multihash": null,
          "providers": null,
          "description": null,
          "instruments": null,
          "end_datetime": null,
          "constellation": null,
          "polarizations": null,
          "start_datetime": null
        }
      },
      "links": [
        {
          "href": "http://localhost:8082/collections/joplin",
          "rel": "collection",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/collections/joplin",
          "rel": "parent",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/",
          "rel": "root",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/collections/joplin/items/ea0fddf4-56f9-4a16-8a0b-f6b0b123b7cf",
          "rel": "self",
          "type": "application/geo+json"
        },
        {
          "href": "http://localhost:8082/collections/joplin/items/ea0fddf4-56f9-4a16-8a0b-f6b0b123b7cf/tiles",
          "rel": "alternate",
          "type": "application/json",
          "title": "tiles"
        }
      ],
      "collection": "joplin"
    },
    {
      "type": "Feature",
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -94.6060181,
              37.0595608
            ],
            [
              -94.6060181,
              37.0332547
            ],
            [
              -94.5730591,
              37.0332547
            ],
            [
              -94.5730591,
              37.0595608
            ],
            [
              -94.6060181,
              37.0595608
            ]
          ]
        ]
      },
      "properties": {
        "gsd": 0.5971642834779395,
        "title": null,
        "width": 2500,
        "height": 2500,
        "created": null,
        "mission": null,
        "updated": null,
        "datetime": "2000-02-02T00:00:00Z",
        "platform": null,
        "proj:epsg": 3857,
        "providers": null,
        "description": null,
        "instruments": null,
        "orientation": "nadir",
        "end_datetime": null,
        "constellation": null,
        "start_datetime": null
      },
      "id": "e0a02e4e-aa0c-412e-8f63-6f5344f829df",
      "bbox": [
        -94.6060181,
        37.0332547,
        -94.5730591,
        37.0595608
      ],
      "stac_version": "1.0.0-beta.2",
      "assets": {
        "COG": {
          "gsd": null,
          "href": "https://arturo-stac-api-test-data.s3.amazonaws.com/joplin/images/may24C357500e4102500n.tif",
          "type": "image/tiff; application=geotiff; profile=cloud-optimized",
          "bands": null,
          "roles": null,
          "title": "NOAA STORM COG",
          "created": null,
          "mission": null,
          "updated": null,
          "platform": null,
          "multihash": null,
          "providers": null,
          "description": null,
          "instruments": null,
          "end_datetime": null,
          "constellation": null,
          "polarizations": null,
          "start_datetime": null
        }
      },
      "links": [
        {
          "href": "http://localhost:8082/collections/joplin",
          "rel": "collection",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/collections/joplin",
          "rel": "parent",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/",
          "rel": "root",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/collections/joplin/items/e0a02e4e-aa0c-412e-8f63-6f5344f829df",
          "rel": "self",
          "type": "application/geo+json"
        },
        {
          "href": "http://localhost:8082/collections/joplin/items/e0a02e4e-aa0c-412e-8f63-6f5344f829df/tiles",
          "rel": "alternate",
          "type": "application/json",
          "title": "tiles"
        }
      ],
      "collection": "joplin"
    },
    {
      "type": "Feature",
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -94.4659424,
              37.1077651
            ],
            [
              -94.4659424,
              37.0814756
            ],
            [
              -94.4329834,
              37.0814756
            ],
            [
              -94.4329834,
              37.1077651
            ],
            [
              -94.4659424,
              37.1077651
            ]
          ]
        ]
      },
      "properties": {
        "gsd": 0.5971642834779395,
        "title": null,
        "width": 2500,
        "height": 2500,
        "created": null,
        "mission": null,
        "updated": null,
        "datetime": "2000-02-02T00:00:00Z",
        "platform": null,
        "proj:epsg": 3857,
        "providers": null,
        "description": null,
        "instruments": null,
        "orientation": "nadir",
        "end_datetime": null,
        "constellation": null,
        "start_datetime": null
      },
      "id": "da6ef938-c58f-4bab-9d4e-89f6ae667da2",
      "bbox": [
        -94.4659424,
        37.0814756,
        -94.4329834,
        37.1077651
      ],
      "stac_version": "1.0.0-beta.2",
      "assets": {
        "COG": {
          "gsd": null,
          "href": "https://arturo-stac-api-test-data.s3.amazonaws.com/joplin/images/may24C370000e4107500n.tif",
          "type": "image/tiff; application=geotiff; profile=cloud-optimized",
          "bands": null,
          "roles": null,
          "title": "NOAA STORM COG",
          "created": null,
          "mission": null,
          "updated": null,
          "platform": null,
          "multihash": null,
          "providers": null,
          "description": null,
          "instruments": null,
          "end_datetime": null,
          "constellation": null,
          "polarizations": null,
          "start_datetime": null
        }
      },
      "links": [
        {
          "href": "http://localhost:8082/collections/joplin",
          "rel": "collection",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/collections/joplin",
          "rel": "parent",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/",
          "rel": "root",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/collections/joplin/items/da6ef938-c58f-4bab-9d4e-89f6ae667da2",
          "rel": "self",
          "type": "application/geo+json"
        },
        {
          "href": "http://localhost:8082/collections/joplin/items/da6ef938-c58f-4bab-9d4e-89f6ae667da2/tiles",
          "rel": "alternate",
          "type": "application/json",
          "title": "tiles"
        }
      ],
      "collection": "joplin"
    },
    {
      "type": "Feature",
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -94.5758057,
              37.0836668
            ],
            [
              -94.5758057,
              37.057369
            ],
            [
              -94.5455933,
              37.057369
            ],
            [
              -94.5455933,
              37.0836668
            ],
            [
              -94.5758057,
              37.0836668
            ]
          ]
        ]
      },
      "properties": {
        "gsd": 0.5971642834779395,
        "title": null,
        "width": 2500,
        "height": 2500,
        "created": null,
        "mission": null,
        "updated": null,
        "datetime": "2000-02-02T00:00:00Z",
        "platform": null,
        "proj:epsg": 3857,
        "providers": null,
        "description": null,
        "instruments": null,
        "orientation": "nadir",
        "end_datetime": null,
        "constellation": null,
        "start_datetime": null
      },
      "id": "d8461d8c-3d2b-4e4e-a931-7ae61ca06dbf",
      "bbox": [
        -94.5758057,
        37.057369,
        -94.5455933,
        37.0836668
      ],
      "stac_version": "1.0.0-beta.2",
      "assets": {
        "COG": {
          "gsd": null,
          "href": "https://arturo-stac-api-test-data.s3.amazonaws.com/joplin/images/may24C360000e4105000n.tif",
          "type": "image/tiff; application=geotiff; profile=cloud-optimized",
          "bands": null,
          "roles": null,
          "title": "NOAA STORM COG",
          "created": null,
          "mission": null,
          "updated": null,
          "platform": null,
          "multihash": null,
          "providers": null,
          "description": null,
          "instruments": null,
          "end_datetime": null,
          "constellation": null,
          "polarizations": null,
          "start_datetime": null
        }
      },
      "links": [
        {
          "href": "http://localhost:8082/collections/joplin",
          "rel": "collection",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/collections/joplin",
          "rel": "parent",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/",
          "rel": "root",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/collections/joplin/items/d8461d8c-3d2b-4e4e-a931-7ae61ca06dbf",
          "rel": "self",
          "type": "application/geo+json"
        },
        {
          "href": "http://localhost:8082/collections/joplin/items/d8461d8c-3d2b-4e4e-a931-7ae61ca06dbf/tiles",
          "rel": "alternate",
          "type": "application/json",
          "title": "tiles"
        }
      ],
      "collection": "joplin"
    },
    {
      "type": "Feature",
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -94.6609497,
              37.1033841
            ],
            [
              -94.6609497,
              37.0770932
            ],
            [
              -94.6279907,
              37.0770932
            ],
            [
              -94.6279907,
              37.1033841
            ],
            [
              -94.6609497,
              37.1033841
            ]
          ]
        ]
      },
      "properties": {
        "gsd": 0.5971642834779395,
        "title": null,
        "width": 2500,
        "height": 2500,
        "created": null,
        "mission": null,
        "updated": null,
        "datetime": "2000-02-02T00:00:00Z",
        "platform": null,
        "proj:epsg": 3857,
        "providers": null,
        "description": null,
        "instruments": null,
        "orientation": "nadir",
        "end_datetime": null,
        "constellation": null,
        "start_datetime": null
      },
      "id": "d4eccfa2-7d77-4624-9e2a-3f59102285bb",
      "bbox": [
        -94.6609497,
        37.0770932,
        -94.6279907,
        37.1033841
      ],
      "stac_version": "1.0.0-beta.2",
      "assets": {
        "COG": {
          "gsd": null,
          "href": "https://arturo-stac-api-test-data.s3.amazonaws.com/joplin/images/may24C352500e4107500n.tif",
          "type": "image/tiff; application=geotiff; profile=cloud-optimized",
          "bands": null,
          "roles": null,
          "title": "NOAA STORM COG",
          "created": null,
          "mission": null,
          "updated": null,
          "platform": null,
          "multihash": null,
          "providers": null,
          "description": null,
          "instruments": null,
          "end_datetime": null,
          "constellation": null,
          "polarizations": null,
          "start_datetime": null
        }
      },
      "links": [
        {
          "href": "http://localhost:8082/collections/joplin",
          "rel": "collection",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/collections/joplin",
          "rel": "parent",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/",
          "rel": "root",
          "type": "application/json"
        },
        {
          "href": "http://localhost:8082/collections/joplin/items/d4eccfa2-7d77-4624-9e2a-3f59102285bb",
          "rel": "self",
          "type": "application/geo+json"
        },
        {
          "href": "http://localhost:8082/collections/joplin/items/d4eccfa2-7d77-4624-9e2a-3f59102285bb/tiles",
          "rel": "alternate",
          "type": "application/json",
          "title": "tiles"
        }
      ],
      "collection": "joplin"
    }
  ],
  "stac_version": "1.0.0-beta.2",
  "links": [
    {
      "href": "http://localhost:8082/collections/joplin/items",
      "rel": "items",
      "type": "application/geo+json"
    },
    {
      "href": "http://localhost:8082/",
      "rel": "parent",
      "type": "application/json"
    },
    {
      "href": "http://localhost:8082/",
      "rel": "root",
      "type": "application/json"
    },
    {
      "href": "http://localhost:8082/collections/joplin",
      "rel": "self",
      "type": "application/json"
    },
    {
      "href": "http://localhost:8082/collections/joplin/items?token=next:d191a6fd-7881-4421-805c-e246371e5cc4",
      "rel": "next",
      "type": "application/json",
      "method": "GET",
      "merge": false
    }
  ],
  "timeStamp": "2021-06-17T14:58:21.337549+00:00"
}  
```
</details>

Reply:
(Unresolving as there's a new issue mentioned, should made a new thread)

Reply:
@lossyrob this was only happening on data ingest, changed the conversion to json used in the pgstac transactions api use exclude_unset=True parameter.

Reply:
> the direct get_landing_page() still needs to be defined for both backends

`get_landing_page` was moved to the base client in #149 so it no longer needs to be defined on the subclasses unless you want different landing page behavior.

Reply:
@geospatial-jeff that won't work here as all_collections is an async function in pgstac. Is there any reason not to move all common function signatures over to using "async def"? It would make it much easier for things like this where we could just use "await all_collections(....)" so that can reuse more code more easily.

Reply:
oh thats a good point.  I don't think exclusively using async (or async) is a good idea.  This is really where having an `AsyncBaseClient` and `BaseClient` would help imo.

Reply:
@geospatial-jeff Looking at the FastAPI docs/recs at https://fastapi.tiangolo.com/async/ it indicates that any non-async end up getting wrapped and awaited in an external threadpool and that it's better to default to using async unless you explicitly need blocking behavior. What are your thoughts on why it may not be good to use async for all path functions? It would certainly be easier / less duplicative to only have a single BaseClient, but I don't really know what any drawbacks would be.

Reply:
@kylebarron you going to just keep thumbsupping contradicting comments? ;-)

Reply:
Calling me out 🙈 Well first I read @geospatial-jeff 's comment, and that made sense, but then I read your comment, and that made sense too 😂

Reply:
Honestly, I claim ignorance on a lot of the intricacies with async & python, so I really have no clue which would be more correct. My laziness definitely tends towards the solution with the least amount of work/repetition.

Reply:
If your route handler is `async def` you need to wrap any synchronous code (ex. sqlalchemy pre v1.4) with a thread pool executor to prevent synchronous calls from blocking the event loop.  This means that a potentially long database call might prevent the fastapi application from handling new requests because the event loop is blocked.  Using a synchronous route handler wraps the entire function call in a thread pool executor which is why it works.  

Also because the backend clients are decoupled from the API layer and may not necessarily be called within the context of FastAPI, we generally don't want `async def` functions that are running blocking code to begin with.

We could switch everything to `async` and just make people aware that they need to be executing synchronous code in threads but I'm not a fan of being opinionated about how code is executed as this will change based on the context.

Reply:
My preference is still creating an `AsyncBaseClient` and `BaseClient` so we can explicitly support both concurrency models.